contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,91 @@
+It does a couple things:
+
+1. Create / Update / (Delete) the resource based on the resources
+2. Only set / delete fields defined in the file either currently or in during last update from the file
+  - Retain updates to fields made by controllers that were never managed in the file
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
+
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
+growing its implementation.  In order to continue to support apply and provide better user visibility
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
+implementation, address gaps in diff-patch semantics preventing correctly applying certain fields,
+and to organize the code in a modular and extendable manner.
+
+## Changes
+
+### Use read-update instead of patch
+
+#### Why
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
+  that will result from the patch.  `--dry-run` is nearly useless for understanding the
+  end state that will result from apply, and implementing a diff operation has been exceedingly
+  copmlex.
+- Generating a patch to make a change is more complicated than simply making the change to
+  the object read from the server.  The apply command reads the live object from the server
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
+  and updating it with a PUT.
+- Patches provide weaker optimistic locking guarantees than PUT.  When encountering an optimistic
+  lock failure in apply, it is easy to re-read the object and try to update it again.  If the
+  apply command tries to modify a field that was modified since apply read the object, it will fail
+  and optionally be able to show this to a user, whereas this conflict may not be detect with a PATCH
+  operation.
+
+#### New approach
+
+- Read the live object
+- Compare the live object to last-recorded and local files
```

Comment:
do you still plan to record last-applied as an annotation?
discussion of the versioning issues (apply v1alpha1, time passes, want to apply v1beta2) would be good as well

Reply:
Yes, this proposal does not change what is stored for any of the objects, and works exactly the same with the exception that it uses PUT instead of PATCH to perform the update.  It might best to solve the issues around versioning in a server-side API.  I imagine the server side API would type convert all 3 objects (recorded from annotation, local from request and remote from live-lookup) to the same version before performing the diff-merge-update.